packages/app/src/components/inspector/PatternInspector.tsx

```diff
@@ -0,0 +1,238 @@
+import Sketch from '@sketch-hq/sketch-file-format-ts';
+import {
+  Select,
+  Spacer,
+  Slider,
+  InputField,
+  getPatternBackground,
+  SketchPattern,
+  imgFileExtensions,
+  mimeTypes,
+  getPatternSize,
+} from 'noya-designsystem';
+import { memo, useCallback, useMemo, useState, DragEvent } from 'react';
+import styled from 'styled-components';
+import * as InspectorPrimitives from './InspectorPrimitives';
+import { fileOpen } from 'browser-fs-access';
+import { uuid } from 'noya-renderer';
+import { FileMap } from 'noya-sketch-file';
+
+const Column = styled.div(({ theme }) => ({
+  flex: '1',
+  display: 'flex',
+  flexDirection: 'column',
+}));
+
+const Container = styled.div<{
+  background: string;
+  backgroundSize: string;
+  repeat: boolean;
+  isDragOver: boolean;
+}>(({ theme, background, backgroundSize, repeat, isDragOver }) => {
+  const dragOverlayColor =
+    theme.colors.primary.replace('rgb', 'rgba').slice(0, -1) + ',0.5)';
+
+  const gradientOverlay = `linear-gradient(0deg, ${dragOverlayColor}, ${dragOverlayColor})`;
+  return {
+    display: 'flex',
+    flex: 1,
+    position: 'relative',
+    outline: 'none',
+    borderRadius: '4px',
+    minHeight: '150px',
+    background: (isDragOver ? `${gradientOverlay},` : '') + background,
+    border: (isDragOver ? '2px ' : '0px ') + theme.colors.primaryDark,
+    backgroundColor: 'white',
+    backgroundPosition: 'center',
+    backgroundRepeat: repeat ? 'auto' : 'no-repeat',
+    backgroundSize,
+    imageRendering: 'crisp-edges',
+    width: '100%',
+    justifyContent: 'center',
+  };
+});
+
+const UploadButton = styled.button<{ show: boolean }>(
+  ({ theme, show = false }) => ({
+    color: theme.colors.button.secondaryText,
+    position: 'absolute',
+    background: 'rgba(0, 0, 0, 0.75)',
+    cursor: 'pointer',
+    userSelect: 'auto',
+    alignSelf: 'center',
+    height: '30px',
+    width: '105px',
+    borderRadius: '24px',
+    border: 'none',
+    display: show ? 'block' : 'none',
+    zIndex: 1,
+  }),
+);
+
+interface Props {
+  id: string;
+  images: FileMap;
+  pattern: SketchPattern;
+  onChangeImage?: (image: Sketch.FileRef | Sketch.DataRef) => void;
```

Comment:
Do these need to be optional? Probably if these callbacks don't exist we shouldn't render this component within the parent at all

Reply:
this is the same issue I have with gradients. All the actions are options because the are optional in the FillPicker. It's there a way of telling it that if the value._class it's gradient and patterns, those functions exist for sure ? 🤔

Reply:
I think so... l'll see if I can come up with something tomorrow morning

Reply:
I would do something similar to the onChanges. I think that way we won't have that many props in some components

Reply:
I tried improving this, but couldn't come up with anything great. The approach you suggested of using `value._class` look like it isn't supported yet but will be soon https://github.com/microsoft/TypeScript/pull/38839

I did see a few other things I want to try refactoring though, after we merge this

Reply:
My brain it's in the future on this one 😆 I'll work on the other changes while I think of a solution for this

Reply:
I work on everything but this part